### PR TITLE
[SYCL][NFC] Refactoring SYCL exception classes.

### PR DESCRIPTION
**Pull request by alexeyvoronov**
- Propagate base classes constructors to the derived exception classes
- Apply clang naming convention to the exception class.
- Deleted useless `#include <exception>`.
- Moved definition `exception::what` method to the library.
- Changed type of `exception::MMsg` field to string_class.
- Removed default value of `exception::MMsg` field.
- Swapped the constructors to avoid creation of extra string_class:
  `exception(const string_class &Msg, ...) : // initialization list {}`
  `exception(const char *Msg, ...) : exception(string_class(Msg)) {}`
- Added missing constructors for classes runtime_error and device_error

Signed-off-by: Alexey Voronov <alexey.voronov@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/exception.hpp [hunk 1]
```diff
@@ -12,40 +12,46 @@
...
-            shared_ptr_class<context> Context = nullptr)
-      : msg(std::string(msg) + " " +
-            ((cl_err == CL_SUCCESS) ? "" : OCL_CODE_TO_STR(cl_err))),
-        cl_err(cl_err), Context(Context) {}
-
-  exception(const std::string &msg, int cl_err = CL_SUCCESS,
-            shared_ptr_class<context> Context = nullptr)
-      : exception(msg.c_str(), cl_err, Context) {}
+  std::string Msg = "Message not specified";
+  cl_int ClErr = CL_SUCCESS;
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/exception.hpp]: Suggest using big l:
ClErr -> CLErr

File: sycl/include/CL/sycl/exception.hpp [hunk 2]
```diff
@@ -12,40 +12,46 @@
...
 struct exception {
+
+  using ContextPtr = shared_ptr_class<context>;
+
   exception() = default;
 
-  const char *what() const noexcept { return msg.c_str(); }
+  exception(const char *Msg, const cl_int ClErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
+      : Msg(std::string(Msg) + " " +
```
Comment (asavonic) [inline sycl/include/CL/sycl/exception.hpp]: You will have a trailing whitespace in case of "ClErr == CL_SUCCESS".

File: sycl/include/CL/sycl/exception.hpp [hunk 3]
```diff
@@ -12,40 +12,46 @@
...
-  exception(const char *msg, int cl_err = CL_SUCCESS,
-            shared_ptr_class<context> Context = nullptr)
-      : msg(std::string(msg) + " " +
-            ((cl_err == CL_SUCCESS) ? "" : OCL_CODE_TO_STR(cl_err))),
-        cl_err(cl_err), Context(Context) {}
-
-  exception(const std::string &msg, int cl_err = CL_SUCCESS,
-            shared_ptr_class<context> Context = nullptr)
-      : exception(msg.c_str(), cl_err, Context) {}
+  std::string Msg = "Message not specified";
```
Comment (asavonic) [inline sycl/include/CL/sycl/exception.hpp]: What is the point of having a default message if you cannot construct an exception class without a user-provided message?

File: sycl/include/CL/sycl/exception.hpp [hunk 4]
```diff
@@ -12,40 +12,46 @@
...
 
-  const char *what() const noexcept { return msg.c_str(); }
+  exception(const char *Msg, const cl_int ClErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
+      : Msg(std::string(Msg) + " " +
+            ((ClErr == CL_SUCCESS) ? "" : OCL_CODE_TO_STR(ClErr))),
+        ClErr(ClErr), Context(Context) {}
+
+  exception(const std::string &Msg, const cl_int ClErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
```
Comment (asavonic) [inline sycl/include/CL/sycl/exception.hpp]: Why do you need a shared pointer to a context, if you only have to return that context by value?

File: sycl/include/CL/sycl/exception.hpp [hunk 5]
```diff
@@ -12,40 +12,46 @@
...
-      : exception(msg.c_str(), cl_err, Context) {}
+  std::string Msg = "Message not specified";
+  cl_int ClErr = CL_SUCCESS;
+  ContextPtr Context;
 };
 
 // Forward declaration
-namespace detail { class queue_impl; }
+namespace detail {
+class queue_impl;
```
Comment (asavonic) [inline sycl/include/CL/sycl/exception.hpp]: Do you need this forward declaration? `friend class detail::queue_impl` below should be a forward declaration itself.

File: sycl/include/CL/sycl/exception.hpp [hunk 6]
```diff
@@ -12,40 +12,46 @@
...
 class context;
 
 struct exception {
+
+  using ContextPtr = shared_ptr_class<context>;
+
   exception() = default;
 
-  const char *what() const noexcept { return msg.c_str(); }
+  exception(const char *Msg, const cl_int ClErr = CL_SUCCESS,
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/exception.hpp]: Could you change semantic of c'tor so it takes usual sycl::context type and store sycl::detail::context_impl inside?

**Comment by bader** [comment]
@alexeyvoronov, ping.

File: sycl/include/CL/sycl/exception.hpp [hunk 7]
```diff
@@ -12,40 +12,48 @@
...
+
+  using ContextPtr = shared_ptr_class<context>;
+
   exception() = default;
 
-  const char *what() const noexcept { return msg.c_str(); }
+  exception(const char *Msg, const cl_int CLErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
+      : Msg(string_class(Msg) + " " +
+            ((CLErr == CL_SUCCESS)
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/exception.hpp]: I don't understand why do you need special handling for CL_SUCCESS, as I can see, it can be correctly handled by `OCL_CODE_TO_STR`

File: sycl/include/CL/sycl/exception.hpp [hunk 8]
```diff
@@ -12,40 +12,48 @@
...
+                 ? string_class(std::to_string(CL_SUCCESS) + " (CL_SUCCESS)")
+                 : OCL_CODE_TO_STR(CLErr))),
+        CLErr(CLErr), Context(Context) {}
+
+  exception(const string_class &Msg, const cl_int CLErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
+      : exception(Msg.c_str(), CLErr, Context) {}
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/exception.hpp]: You are getting c-string here to pass it to the next constructor which will again create `string_class` from it.
I think it would be better to "swap" constructors:

```c++
exception(const string_class &Msg, ...) : // some code {}
exception(const chat *Msg, ...) : exception(string_class(Msg))
```

Comment (alexeyvoronov) [reply sycl/include/CL/sycl/exception.hpp]: Done.

File: sycl/include/CL/sycl/exception.hpp (hunk 7 above)
Comment (alexeyvoronov) [reply sycl/include/CL/sycl/exception.hpp]: Done.

File: sycl/include/CL/sycl/detail/queue_impl.hpp [hunk 9]
```diff
@@ -53,6 +54,7 @@ class queue_impl {
   }
 
   ~queue_impl() {
+    throw_asynchronous();
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/detail/queue_impl.hpp]: This doesn't look as a refactoring. I think it is better to move this change at least to a separate commit

File: sycl/include/CL/sycl/exception.hpp (hunk 4 above)
Comment (asavonic) [reply sycl/include/CL/sycl/exception.hpp]: None of these seems like an argument: using a shared ptr accompanied by a heap allocation is a huge overhead _just_ to get an additional state (context is missing). One extra bool variable should do the trick.

File: sycl/include/CL/sycl/detail/queue_impl.hpp (hunk 9 above)
Comment (alexeyvoronov) [reply sycl/include/CL/sycl/detail/queue_impl.hpp]: Done.

File: sycl/include/CL/sycl/exception.hpp (hunk 4 above)
Comment (alexeyvoronov) [reply sycl/include/CL/sycl/exception.hpp]: Will be done in a separate patch. This patch I will leave the as NFC.

File: sycl/include/CL/sycl/exception.hpp [hunk 10]
```diff
@@ -73,36 +77,57 @@ using async_handler = function_class<void(cl::sycl::exception_list)>;
...
-  runtime_error(const char *str, cl_int err = CL_SUCCESS)
-      : exception(str, err) {}
-  runtime_error(const std::string &str, cl_int err = CL_SUCCESS)
-      : runtime_error(str.c_str(), err) {}
+  runtime_error() = default;
+
+  runtime_error(const char *Msg, cl_int Err = CL_SUCCESS)
+      : runtime_error(string_class(Msg), Err) {}
+
+  runtime_error(const string_class &Msg, cl_int Err = CL_SUCCESS)
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/exception.hpp]: Probably you need to use pi counterparts here.

Comment (alexeyvoronov) [reply sycl/include/CL/sycl/exception.hpp]: Yes, in the future I will make it.

File: sycl/include/CL/sycl/exception.hpp [hunk 11]
```diff
@@ -12,40 +12,45 @@
...
 namespace sycl {
 
+// Forward declaration
 class context;
 
 struct exception {
   exception() = default;
 
-  const char *what() const noexcept { return MMsg.c_str(); }
+  const char *what() const noexcept;
```
Comment (asavonic) [inline sycl/include/CL/sycl/exception.hpp]: Why you want to move this function to the library?

File: sycl/include/CL/sycl/exception.hpp [hunk 12]
```diff
@@ -36,15 +36,14 @@ struct exception {
...
-      : MMsg(string_class(Msg) + " " +
-             ((CLErr == CL_SUCCESS) ? "" : OCL_CODE_TO_STR(CLErr))),
-        MCLErr(CLErr), MContext(Context) {}
+      : exception(string_class(Msg), CLErr, Context) {}
 
-  exception(const string_class &Msg, int CLErr = CL_SUCCESS,
+  exception(const string_class &Msg, const cl_int CLErr = CL_SUCCESS,
             shared_ptr_class<context> Context = nullptr)
-      : exception(Msg.c_str(), CLErr, Context) {}
+      : MMsg(Msg + " " + OCL_CODE_TO_STR(CLErr)), MCLErr(CLErr),
```
Comment (asavonic) [inline sycl/include/CL/sycl/exception.hpp]: If `CLErr == CL_SUCCESS`, the message is going to look weird.

File: sycl/include/CL/sycl/exception.hpp (hunk 11 above)
Comment (alexeyvoronov) [reply sycl/include/CL/sycl/exception.hpp]: Because other three functions in the library.
So why not?

File: sycl/include/CL/sycl/exception.hpp (hunk 12 above)
Comment (alexeyvoronov) [reply sycl/include/CL/sycl/exception.hpp]: Hm, do you agree with it?
`MMsg(Msg + ((CLErr == CL_SUCCESS) ? "" :  " " + OCL_CODE_TO_STR(CLErr)))`

Comment (asavonic) [reply sycl/include/CL/sycl/exception.hpp]: That should work. You probably need to wrap `" " + OCL_CODE_TO_STR(CLErr)` into an `std::string`, because you cannot concatenate C strings.
